operators/pkg/controller/elasticsearch/driver/upgrade.go

```diff
@@ -0,0 +1,301 @@
+				//  - zen2: set voting config exclusions: same problem, this is not easy. But since we only delete
+				//   one master at a time, maybe it's not required?
+			}
+
+			// Upgrade the pod.
+			if err := d.upgradeStatefulSetPartition(&statefulSets[i], partition); err != nil {
+				return results.WithError(err)
+			}
+			scheduledUpgrades++
+		}
+	}
+	return results
+}
+
+func (d *defaultDriver) upgradeStatefulSetPartition(
+	statefulSet *appsv1.StatefulSet,
+	newPartition int32,
+) error {
+	// Node can be removed, update the StatefulSet rollingUpdate.Partition ordinal.
+	log.Info("Updating rollingUpdate.Partition",
+		"namespace", statefulSet.Namespace,
+		"name", statefulSet.Name,
+		"from", statefulSet.Spec.UpdateStrategy.RollingUpdate.Partition,
+		"to", &newPartition,
+	)
+	statefulSet.Spec.UpdateStrategy.RollingUpdate = &appsv1.RollingUpdateStatefulSetStrategy{
+		Partition: &newPartition,
+	}
+	if err := d.Client.Update(statefulSet); err != nil {
+		return err
+	}
+
+	// Register the updated sset generation to deal with out-of-date sset cache.
+	d.Expectations.ExpectGeneration(statefulSet.ObjectMeta)
+
+	return nil
+}
+
+func prepareClusterForNodeRestart(esClient esclient.Client, esState ESState) error {
+	// Disable shard allocations to avoid shards moving around while the node is temporarily down
+	shardsAllocationEnabled, err := esState.ShardAllocationsEnabled()
+	if err != nil {
+		return err
+	}
+	if shardsAllocationEnabled {
+		if err := disableShardsAllocation(esClient); err != nil {
+			return err
+		}
+	}
+
+	// Request a sync flush to optimize indices recovery when the node restarts.
+	if err := doSyncFlush(esClient); err != nil {
+		return err
+	}
+
+	// TODO: halt ML jobs on that node
+	return nil
+}
+
+// clusterReadyForNodeRestart returns true if the ES cluster allows a node to be restarted
+// with minimized downtime and no unexpected data loss.
+func clusterReadyForNodeRestart(es v1alpha1.Elasticsearch, esState ESState) (bool, error) {
+	// Check the cluster health: only allow node restart if health is green.
+	// This would cause downtime if some shards have 0 replicas, but we consider that's on the user.
+	// TODO: we could technically still restart a node if the cluster is yellow,
+	//  as long as there are other copies of the shards in-sync on other nodes
+	// TODO: the fact we rely on a cached health here would prevent more than 1 restart
+	//  in a single reconciliation
+	green, err := esState.GreenHealth()
+	if err != nil {
+		return false, err
+	}
+	if !green {
+		log.Info("Skipping node rolling upgrade since cluster is not green", "namespace", es.Namespace, "name", es.Name)
+		return false, nil
+	}
+	return true, nil
+}
+
+// podUpgradeDone inspects the given pod and returns true if it was successfully upgraded.
+func podUpgradeDone(c k8s.Client, esState ESState, podRef types.NamespacedName, expectedRevision string) (bool, error) {
+	if expectedRevision == "" {
```

Comment:
Should this be checked to skip the entire nodeset ? After this we proceed with doRollingUpgrade, aren't we?.

Reply:
Yes, we could. That would be a small optimization, but it would split the upgrade check condition logic to 2 different places, so not sure it's worth it?

Reply:
I see, I agree. Follow up question: why not skip this check here and instead filter ssets for upgrade in RevisionUpdateScheduled?

Reply:
You're right. It's actually already checked there, and we skip the entire upgrade if there is no update revision: 

```go
func (l StatefulSetList) RevisionUpdateScheduled() bool {
	for _, s := range l {
		if s.Status.UpdateRevision != "" && s.Status.UpdateRevision != s.Status.CurrentRevision {
			return true
		}
	}
	return false
}
```

I still think I'd prefer to keep the (useless) check in `podUpgradeDone` though. It makes this small function "complete" in the sense that it should catch all corner cases. If at some point we shuffle things around that's a corner case we need not to forget. Similar to eg. checking `nil` values in some functions where, if we correctly looked several layers above, we would know the value can never be `nil`. WDYT?

Reply:
I think it's fine to leave to check. Would you change RevisionUpdateScheduled and run the loop only on ssets that need the upgrade or leave it as it is?

Reply:
I can probably do something like:

```go
for i, statefulSet := range statefulSets.ToUpdate() {
```

instead of:

```go
if !statefulSets.RevisionUpdateScheduled() {
		// nothing to upgrade
		return results
	}

(...)
for i, statefulSet := range statefulSets {
```

Reply:
Yep, that's what I was thinking too.